File: apps/frontend/src/lib/schemas/evidence.ts
import { z } from 'zod';

import {
  confidenceLabelSchema,
  entityTypeSchema,
  evidenceKindSchema,
  evidenceStanceSchema,
  hypothesisStatusSchema,
  isoDateTimeSchema,
  modalitySchema,
  provenanceSchema,
} from './primitives';

export const evidenceItemSchema = z.object({
  evidence_id: z.string(),
  kind: evidenceKindSchema,
  modality: modalitySchema,
  summary: z.string().default(''),
  excerpt: z.string().default(''),
  provenance: provenanceSchema,
  stance: evidenceStanceSchema.default('neutral'),
  relevance: z.number().default(0),
  reliability: z.number().default(0.5),
  reliability_reason: z.string().default(''),
  /** Present on `GET /api/investigations/{id}`; the API renders `provenance.human()`. */
  citation: z.string().optional(),
  entities: z.array(z.string()).default([]),
  occurred_at: isoDateTimeSchema.nullable().optional(),
  retrieved_at: isoDateTimeSchema.optional(),
  retrieved_by: z.string().default(''),
  hypothesis_ids: z.array(z.string()).default([]),
});
export type EvidenceItem = z.infer<typeof evidenceItemSchema>;

export const hypothesisSchema = z.object({
  hypothesis_id: z.string(),
  description: z.string(),
  rationale: z.string().default(''),
  status: hypothesisStatusSchema.default('open'),
  confidence: z.number().default(0),
  prior: z.number().default(0.25),
  supporting_evidence: z.array(z.string()).default([]),
  contradicting_evidence: z.array(z.string()).default([]),
  disproof_probe: z.string().nullable().default(null),
  disproof_searched: z.boolean().default(false),
  verified: z.boolean().default(false),
  verification_note: z.string().default(''),
  predicted_signals: z.array(z.string()).default([]),
});
export type Hypothesis = z.infer<typeof hypothesisSchema>;

export const contradictionSchema = z.object({
  contradiction_id: z.string(),
  statement: z.string(),
  evidence_a: z.string(),
  evidence_b: z.string(),
  kind: z.string().default('value_conflict'),
  detail: z.string().default(''),
  resolution: z.string().nullable().default(null),
  resolved_in_favour_of: z.string().nullable().default(null),
  severity: z.number().default(0.5),
  entity_id: z.string().nullable().default(null),
});
export type Contradiction = z.infer<typeof contradictionSchema>;

export const timelineEventSchema = z.object({
  event_id: z.string(),
  occurred_at: isoDateTimeSchema,
  label: z.string(),
  detail: z.string().default(''),
  modality: modalitySchema,
  evidence_ids: z.array(z.string()).default([]),
  entity_ids: z.array(z.string()).default([]),
  source_id: z.string().nullable().default(null),
  precision: z.string().default('exact'),
});
export type TimelineEvent = z.infer<typeof timelineEventSchema>;

export const claimSchema = z.object({
  claim_id: z.string(),
  text: z.string(),
  confidence: z.number(),
  status: z.string().default('supported'),
  evidence_ids: z.array(z.string()).default([]),
});
export type Claim = z.infer<typeof claimSchema>;

export const applicationLinkSchema = z.object({
  entity_id: z.string(),
  entity_type: entityTypeSchema,
  label: z.string(),
  url: z.string(),
  record_id: z.string(),
  verified_in_database: z.boolean().default(false),
});
export type ApplicationLink = z.infer<typeof applicationLinkSchema>;

export const verificationResultSchema = z.object({
  claim: z.string(),
  supported: z.boolean(),
  confidence: z.number(),
  supporting_evidence: z.array(z.string()).default([]),
  contradicting_evidence: z.array(z.string()).default([]),
  diversity: z.number().default(0),
  note: z.string().default(''),
  checks: z.record(z.boolean()).default({}),
});
export type VerificationResult = z.infer<typeof verificationResultSchema>;

export const canonicalEntitySchema = z.object({
  entity_id: z.string(),
  entity_type: entityTypeSchema,
  canonical_name: z.string(),
  aliases: z.array(z.string()).default([]),
  normalized_keys: z.array(z.string()).default([]),
  confidence: z.number().default(1),
  source_ids: z.array(z.string()).default([]),
  modalities: z.array(modalitySchema).default([]),
  attributes: z.record(z.unknown()).default({}),
  mention_count: z.number().default(0),
  first_seen: isoDateTimeSchema.optional(),
  last_seen: isoDateTimeSchema.optional(),
});
export type CanonicalEntity = z.infer<typeof canonicalEntitySchema>;

export const resolutionCandidateSchema = z.object({
  entity_id: z.string(),
  canonical_name: z.string(),
  entity_type: entityTypeSchema,
  score: z.number(),
  method: z.string(),
  rationale: z.string().nullable().optional(),
});

export const entityResolutionSchema = z.object({
  query: z.string(),
  normalized: z.string(),
  resolved: canonicalEntitySchema.nullable().default(null),
  confidence: z.number().default(0),
  method: z.string().default('none'),
  candidates: z.array(resolutionCandidateSchema).default([]),
  needs_verification: z.boolean().default(false),
  explanation: z.string().default(''),
});
export type EntityResolution = z.infer<typeof entityResolutionSchema>;

/**
 * `GET /api/entities/{id}` adds `cross_modal_links` and `application_links` to the
 * canonical entity. `docs/api.md` names both but does not give their shape, so the
 * link rows are parsed permissively and every field is treated as optional.
 */
export const crossModalLinkSchema = z.object({
  modality: modalitySchema.optional(),
  asset_id: z.string().optional(),
  chunk_id: z.string().optional(),
  source_id: z.string().optional(),
  title: z.string().nullable().optional(),
  surface: z.string().optional(),
  confidence: z.number().optional(),
});
export type CrossModalLink = z.infer<typeof crossModalLinkSchema>;

export const entityDetailSchema = canonicalEntitySchema.extend({
  cross_modal_links: z.array(crossModalLinkSchema).default([]),
  application_links: z.array(applicationLinkSchema).default([]),
});
export type EntityDetail = z.infer<typeof entityDetailSchema>;

export const graphNodeSchema = z.object({
  node_id: z.string(),
  labels: z.array(z.string()).default([]),
  properties: z.record(z.unknown()).default({}),
});
export type GraphNodeData = z.infer<typeof graphNodeSchema>;

export const graphEdgeSchema = z.object({
  edge_id: z.string(),
  type: z.string(),
  start: z.string(),
  end: z.string(),
  properties: z.record(z.unknown()).default({}),
});
export type GraphEdgeData = z.infer<typeof graphEdgeSchema>;

export const graphViewSchema = z.object({
  nodes: z.array(graphNodeSchema).default([]),
  edges: z.array(graphEdgeSchema).default([]),
  truncated: z.boolean().default(false),
});
export type GraphView = z.infer<typeof graphViewSchema>;

export { confidenceLabelSchema };

File: packages/schemas/spectra_schemas/evidence.py
"""Evidence ledger, claims, contradictions, timeline and verification."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .enums import (
    ClaimStatus,
    ConfidenceLabel,
    EntityType,
    EvidenceKind,
    EvidenceStance,
    Modality,
)
from .provenance import Provenance


def _now() -> datetime:
    return datetime.now(timezone.utc)


class EvidenceItem(BaseModel):
    """One citable fact with its full provenance and reliability accounting.

    Nothing reaches the user as fact without a corresponding EvidenceItem.
    """

    model_config = ConfigDict(extra="forbid")

    evidence_id: str
    kind: EvidenceKind
    modality: Modality
    summary: str = Field(description="Short extractive statement of what this evidence shows")
    excerpt: str = ""
    provenance: Provenance
    stance: EvidenceStance = EvidenceStance.NEUTRAL
    relevance: float = 0.0
    reliability: float = 0.5
    reliability_reason: str = ""
    entities: list[str] = Field(default_factory=list)
    occurred_at: datetime | None = None
    retrieved_at: datetime = Field(default_factory=_now)
    retrieved_by: str = ""
    claim_ids: list[str] = Field(default_factory=list)

    @property
    def weight(self) -> float:
        """Combined usefulness: how relevant AND how trustworthy."""
        return round(self.relevance * self.reliability, 6)

    def citation(self) -> str:
        return self.provenance.human()


class EvidenceLedger(BaseModel):
    """Append-only record of everything considered, used and rejected."""

    model_config = ConfigDict(extra="forbid")

    items: list[EvidenceItem] = Field(default_factory=list)
    rejected_count: int = 0
    considered_count: int = 0
    rejection_reasons: dict[str, int] = Field(default_factory=dict)

    def with_items(self, new_items: list[EvidenceItem]) -> EvidenceLedger:
        """Return a new ledger - ledgers are never mutated in place."""
        known = {i.evidence_id for i in self.items}
        merged = list(self.items) + [i for i in new_items if i.evidence_id not in known]
        return EvidenceLedger(
            items=merged,
            rejected_count=self.rejected_count,
            considered_count=self.considered_count + len(new_items),
            rejection_reasons=dict(self.rejection_reasons),
        )

    def diversity(self) -> float:
        """Independent-corroboration score in [0,1].

        Three paragraphs of one PDF must not outrank a DB record + a PDF + a
        video, so diversity counts distinct sources AND distinct modalities.
        """
        if not self.items:
            return 0.0
        sources = {i.provenance.source_id for i in self.items}
        modalities = {i.modality for i in self.items}
        assets = {_asset_of(i) for i in self.items}
        source_term = min(len(sources) / 3.0, 1.0)
        modality_term = min(len(modalities) / 3.0, 1.0)
        asset_term = min(len(assets) / 4.0, 1.0)
        return round(0.4 * source_term + 0.4 * modality_term + 0.2 * asset_term, 4)

    def by_stance(self, stance: EvidenceStance) -> list[EvidenceItem]:
        return [i for i in self.items if i.stance is stance]


def _asset_of(item: EvidenceItem) -> str:
    loc = item.provenance.locator
    for attr in ("document_id", "image_id", "video_id", "audio_id", "record_id", "resource"):
        value = getattr(loc, attr, None)
        if value:
            return str(value)
    return item.evidence_id


class Contradiction(BaseModel):
    """Two pieces of evidence that cannot both be true."""

    model_config = ConfigDict(extra="forbid")

    contradiction_id: str
    statement: str
    evidence_a: str
    evidence_b: str
    kind: str = "value_conflict"
    detail: str = ""
    resolution: str | None = None
    resolved_in_favour_of: str | None = None
    severity: float = 0.5
    entity_id: str | None = None


class TimelineEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_id: str
    occurred_at: datetime
    label: str
    detail: str = ""
    modality: Modality
    evidence_ids: list[str] = Field(default_factory=list)
    entity_ids: list[str] = Field(default_factory=list)
    source_id: str | None = None
    precision: str = "exact"


class VerificationResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    claim: str
    supported: bool
    confidence: float
    supporting_evidence: list[str] = Field(default_factory=list)
    contradicting_evidence: list[str] = Field(default_factory=list)
    diversity: float = 0.0
    note: str = ""
    checks: dict[str, bool] = Field(default_factory=dict)


class Claim(BaseModel):
    """A single asserted statement, judged on the evidence that backs it.

    Each claim stands or falls on its own evidence. There is no competition
    between rival explanations, so a well-supported conclusion keeps its
    confidence instead of having it divided among alternatives.
    """

    model_config = ConfigDict(extra="forbid")

    claim_id: str
    text: str
    confidence: float = 0.0
    status: ClaimStatus = ClaimStatus.INSUFFICIENT
    supporting_evidence: list[str] = Field(default_factory=list)
    contradicting_evidence: list[str] = Field(default_factory=list)
    disproof_probe: str | None = Field(
        default=None, description="What evidence would show this claim is wrong"
    )
    disproof_searched: bool = False
    verified: bool = False
    verification_note: str = ""
    rationale: str = ""

    @property
    def evidence_ids(self) -> list[str]:
        """Every piece of evidence bearing on this claim, either way."""
        return [*self.supporting_evidence, *self.contradicting_evidence]

    @property
    def confidence_label(self) -> ConfidenceLabel:
        return ConfidenceLabel.from_score(self.confidence)


class ApplicationLink(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entity_id: str
    entity_type: EntityType
    label: str
    url: str
    record_id: str
    verified_in_database: bool = False


class GraphNode(BaseModel):
    node_id: str
    labels: list[str] = Field(default_factory=list)
    properties: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    edge_id: str
    type: str
    start: str
    end: str
    properties: dict[str, Any] = Field(default_factory=dict)


class GraphView(BaseModel):
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    truncated: bool = False
